src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java

```diff
@@ -634,12 +666,27 @@ private void setKeyEntry(String alias, Key key,
                 secretKeyInfo.putInteger(0);
                 AlgorithmId algId = AlgorithmId.get(key.getAlgorithm());
                 algId.encode(secretKeyInfo);
-                secretKeyInfo.putOctetString(key.getEncoded());
+                byte[] encoded = key.getEncoded();
+                // Write zeroes of correct length into intermediate
+                // DerOutputStream so that the real content will not
+                // appear everywhere. The actual bytes will be copied
+                // into the final encoding so they can be cleared
+                // effectively once required.
+                secretKeyInfo.putOctetString(new byte[encoded.length]);
                 pkcs8.write(DerValue.tag_Sequence, secretKeyInfo);
```

Comment:
Same as earlier comment - can switch to DerValue.wrap() and then clear() approach?